Part 7 - Zarr consolidated metadata

mannreis opened the pull request:
This PR follows up on #3224 :
 - Adds `mode` control parameters `csl`,`consolidate`,`consolidated` to select the use of consolitated metadata.
 - Adds directives to control the via environment varibles the use of consolidated metadata.
 - Modifies the `NCZMD_set_metadata_handler` to select the handler based on the settings, and presence of consolidated metadata.
 - Extends the tests to confirm the correct usage of consolidated metadata

mannreis committed:
Expose ncstrndup as strndup to whole libnetcdf

mannreis committed:
include: define NOTZARR and ZARRMETA errors

mannreis committed:
zmetadata: introduce header

mannreis committed:
zmetadata: add consolidated implementation

mannreis committed:
zmetadata: introduce wrappers and implementation

mannreis committed:
Document zmetadata.h

mannreis committed:
zmetadata: document consolidated implementation

mannreis committed:
Document zmetadata2.c

mannreis committed:
zarr: set metadata handler on ncz_open_dataset

mannreis committed:
rename Z* directives to Z2*

mannreis committed:
zarr: define FLAG_CONSOLIDATED set via mode=csl|consolidate|consolidated

mannreis committed:
zclose: reclaim non consolidated JSON

mannreis committed:
zsync.c: rename fullpath to key

mannreis committed:
zsync: replace `NCZ_upload` and `updload_attrs`, etc with metadata handler `NCZMD_upload_json_*`

mannreis committed:
zsync.c: rename varpath to key

mannreis committed:
zsync: replace `downloadzarrjson` with metadata handler `fetch_json_*`

mannreis committed:
zarr: set metadata handler on ncz_create_dataset

mannreis committed:
tests: extend run_newformat.sh with consolidated metadata case

mannreis committed:
zsync.c: replace search_vars and search_nodes with metadata handler `list_nodes-*`

mannreis committed:
zsync.c: call NCZMD_consolidate when syncing group/file

mannreis committed:
zarr.*: NCZ_uploadjson, set `const NCjson*`

mannreis committed:
zmetadata.h: directives to select the use of consolidated metadata via environment variables

mannreis committed:
zmetadata.c: change NCZMD_set_metadata_handler to take setting into account

mannreis committed:
zmetadata.h: Define `NCZMD_consolidate`

mannreis committed:
Address strndup review

DennisHeimbigner commented:
Ok, I have reviewed my initial notes are as follows:

* csl-1-5 seem ok.
* csl-6:
    - in NCZMD_v2_csl_list_nodes, you appear to be trying to parse  keys yourself. There are some key split/join functions (see include ncutil.h) that might be useful here.
    - Otherwise ok.

* csl-7:
    - using three different mode keys -- csl,consolidate,consolidated -- to signal  consolidated format is probably excessive. You should probably settle on one.
    - otherwise ok

DennisHeimbigner committed:
Expose NCZMD_consolidate for use in zsync.c

DennisHeimbigner committed:
Expose NCZMD_consolidate for use in zsync.c

DennisHeimbigner committed:
structify

DennisHeimbigner committed:
typo

DennisHeimbigner committed:
tracking down error

DennisHeimbigner committed:
ugly

mannreis committed:
NCZMD_v2_csl_list_nodes: Address review sugestion to use NC_split_delim

mannreis committed:
zsync.c: fix memory leak

mannreis committed:
zarr: address review `mode=consolidated`

mannreis committed:
tests: test `mode` controls for `consolidated`

mannreis commented:
@DennisHeimbigner I've force pushed to this branch to align all 7 PRs! Please `pull --rebase` if you intend to add anything.

I'll just go over the overall feedback:
> * csl-1-5 seem ok.

I've fixed a memory leak via 05b4732

> * csl-6:
>   
>   * in NCZMD_v2_csl_list_nodes, you appear to be trying to parse  keys yourself. There are some key split/join functions (see include ncutil.h) that might be useful here.

0b8a548 addresses this by using `NC_split_delim`

> * csl-7:
>   
>   * using three different mode keys -- csl,consolidate,consolidated -- to signal  consolidated format is probably excessive. You should probably settle on one.

a48f9d4 and 7d1b361 address your last suggestion to keep only `mode=consolidated`

>   * This PR did not pass all tests, so I fixed it.

You can see now that my commit b0b74a51 (in Part 5 zarr-csl-5) is equalent to what you added here in part 7 (e938fbd + ff3d0f9). So we can clean this *after* part 5 or 6 is are merged.

Many thanks for your prompt review! And I'm sorry for this `git` puzzle

CC: @WardF

mannreis commented:
Ok! Then I'd appreciate it if you could give your final approval to the PRs. Thanks!

mannreis commented:
Thanks a lot for the speedy review! Take your time for  the final PR and don't hesitate if you need more clarifications/discussion.
